tls/extensions/s2n_client_early_data_indication.c

```diff
@@ -0,0 +1,140 @@
+ */
+
+#include <s2n.h>
+
+#include "tls/extensions/s2n_early_data_indication.h"
+
+#include "tls/extensions/s2n_client_psk.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_early_data.h"
+#include "tls/s2n_protocol_preferences.h"
+#include "tls/s2n_tls13.h"
+#include "utils/s2n_safety.h"
+
+static S2N_RESULT s2n_early_data_config_is_possible(struct s2n_connection *conn)
+{
+    ENSURE_REF(conn);
+
+    struct s2n_psk *first_psk = NULL;
+    GUARD_RESULT(s2n_array_get(&conn->psk_params.psk_list, 0, (void**) &first_psk));
+    ENSURE_REF(first_psk);
+
+    struct s2n_early_data_config *config = &first_psk->early_data_config;
+
+    /* Must support early data */
+    ENSURE_GT(config->max_early_data_size, 0);
+
+    /* Early data must require a protocol than we could negotiate */
+    ENSURE_GTE(s2n_connection_get_protocol_version(conn), config->protocol_version);
+    ENSURE_GTE(s2n_connection_get_protocol_version(conn), S2N_TLS13);
+
+    const struct s2n_cipher_preferences *cipher_preferences = NULL;
+    GUARD_AS_RESULT(s2n_connection_get_cipher_preferences(conn, &cipher_preferences));
+    ENSURE_REF(cipher_preferences);
+
+    /* Early data must require a supported cipher */
+    bool match = false;
+    for (uint8_t i = 0; i < cipher_preferences->count; i++) {
+        if (memcmp(cipher_preferences->suites[i]->iana_value, config->cipher_suite_iana, S2N_TLS_CIPHER_SUITE_LEN) == 0) {
+            match = true;
+            break;
+        }
+    }
+    ENSURE_EQ(match, true);
+
+    /* If early data specifies an application protocol, it must be supported by protocol preferences */
+    if (config->application_protocol.size > 0) {
+        struct s2n_blob *application_protocols = NULL;
+        GUARD_AS_RESULT(s2n_connection_get_protocol_preferences(conn, &application_protocols));
+        ENSURE_REF(application_protocols);
+
+        match = false;
+        GUARD_RESULT(s2n_protocol_preferences_contain(application_protocols, &config->application_protocol, &match));
+        ENSURE_EQ(match, true);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static bool s2n_client_early_data_indication_should_send(struct s2n_connection *conn)
+{
+    return s2n_result_is_ok(s2n_early_data_config_is_possible(conn))
+            /**
+             *= https://tools.ietf.org/rfc/rfc8446#section-4.2.10
+             *# A client MUST NOT include the
+             *# "early_data" extension in its followup ClientHello.
+             **/
+            && !s2n_is_hello_retry_handshake(conn)
+            /**
+             *= https://tools.ietf.org/rfc/rfc8446#section-4.2.10
+             *# When a PSK is used and early data is allowed for that PSK, the client
+             *# can send Application Data in its first flight of messages.  If the
+             *# client opts to do so, it MUST supply both the "pre_shared_key" and
+             *# "early_data" extensions.
+             */
+            && s2n_client_psk_extension.should_send(conn);
```

Comment:
`return conn && s2n_result_is_ok(s2n_early_data_config_is_possible(conn))`

Reply:
Why? We're not dereferencing conn here.